Optimize things a little bit

gaearon opened the pull request:
I updated the code to be more idiomatic Redux. I think it performs a little better in my testing.
I might be able to take another look tomorrow evening in case there’s something else I could optimize.

The basic idea here (apart from minor refactorings) is to separate independently changing data (the list of items and items themselves). This, of course, highly depends on the specific app, which is why we don’t have a proper guide for doing it, but the point is that it’s usually possible to optimize an app for specific use cases if there is a need for it.

[Computing Derived Data](http://redux.js.org/docs/recipes/ComputingDerivedData.html) is probably the best we have in terms of guides on how to do it, but it definitely could use some state structure refactoring advice content.

cc @ellbee for no specific reason—just in case he’s interested

gaearon committed:
Optimize things a little bit

gaearon committed:
Fix a bug

gaearon committed:
Separate components more optimally

gaearon committed:
Stylistic tweaks

capaj commented:
This is freaking awesome. I can't wait to see the updated charts @mweststrate!

Kerumen commented:
Just wondering why you add a `shouldComponentUpdate` even if the component is `connect`ed?
If I read correctly the [react-redux docs](https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments), you can specifiy if a component implements a  `shouldComponentUpdate` with the fourth argument of `connect`. The default is `true`. So by default, all components `connect`ed are pure, there is no need to write it.

No doubt [you are aware](https://twitter.com/dan_abramov/status/719723707124604928?s=09) of this, I didn't understand it correctly I guess.

mweststrate commented:
@Kerumen you mean like in the original? For the first scenario (without selectors) `TodoItem` didn't use `connect` at all but just received it's todo item from the parent. So not using PureRenderMixin would have caused all `TodoItem` components to re-render if a todo changes, instead of only the affected ones.

Kerumen commented:
No I was asking for the @gaearon rewrite. For instance [here](https://github.com/mweststrate/redux-todomvc/pull/1/commits/3ff90c9f1d695468b3de8c4aa80a4b2fdf9aeea7?diff=unified#diff-7529e3d566b5117b513a5da85f5d4d53R15).

michaseel commented:
@gaearon Could you update the unit tests too? I would like to see how you test stateless functional components!

gaearon commented:
Well, it’s not _that_ much of an improvement but it’s definitely better. Still can’t avoid going through 10K subscribers. There may be small things to optimize in React Redux as well so I’ll look there. 

I removed `shouldComponentUpdate` later, please see the complete change list. I added it for testing in one of the commits but decided it is unnecessary.

No, alas, I won’t be adding tests here, you should check out Redux repo (and open PRs to it) to see how to test components.

slorber commented:
> Still can’t avoid going through 10K subscribers.

I guess 10k subscribers is better than 10k items list reconciliation.

But there are potential ways to avoid calling those 10k subscribers on every item change. Didn't have time to work on something concrete yet but it can be optimized further imho. See https://github.com/reactjs/react-redux/issues/269

gaearon commented:
Yeah, there are definitely ways to solve this even with vanilla Redux. The goal here is to avoid complicating the code beyond recognition 🙂

gaearon committed:
Use perf improvements in react-redux@4.4.4

narqo commented on components/App.js:
```diff
@@ -0,0 +1,13 @@
+import React, { PropTypes } from 'react'
+import Header from '../components/Header'
+import MainSection from '../components/MainSection'
+import * as TodoActions from '../actions'
```
This `TodoActions` import is useless here, isn't it?

gaearon replied:
Yea, missed this one. Doesn't affect the perf though.

mweststrate commented:
Thanks for the contributions! Will re-run benchmarks tomorrow.

gaearon committed:
Uglify in prod

gaearon committed:
Minor fix
